bns-test/test/namespace_reveal.test.ts

```diff
@@ -0,0 +1,193 @@
+      expect(receipt.success).eq(false);
+      expect(receipt.error).include('1001');
+    });
+
+    it("should fail if TTL expired", async () => {
+      await mineBlocks(bns, 11);
+      let receipt = await bns.namespaceReveal(
+        cases[0].namespace, 
+        cases[0].salt,
+        cases[0].priceFunction, 
+        cases[0].renewalRule, 
+        cases[0].nameImporter, { sender: bob });
+      expect(receipt.success).eq(false);
+      expect(receipt.error).include('1001');
+    });
+  });
+
+  describe("Given an existing, valid pre-order from Alice for the namespace 'blockstack' initiated at block #31, revealing the namespace", () => {
+
+    beforeEach(async () => {
+      let receipt = await bns.namespacePreorder(cases[0].namespace, cases[0].salt, cases[0].value, { sender: cases[0].namespaceOwner });
+      expect(receipt.success).eq(true);
+      expect(receipt.result).include('u12');
+    });
+
+
+    it("should succeed if the price-function, lifetime, namespace and salt are valid", async () => {
+      let receipt = await bns.namespaceReveal(
+        cases[0].namespace, 
+        cases[0].salt,
+        cases[0].priceFunction, 
+        cases[0].renewalRule, 
+        cases[0].nameImporter, { sender: cases[0].namespaceOwner });
+      expect(receipt.success).eq(true);
+      expect(receipt.result).include('true');
+    });
+  });
+
+  describe("Given a pre-order, too cheap, from Bob for the namespace 'id' initiated at block #31, revealing the namespace", () => {
+
+    beforeEach(async () => {
+      let receipt = await bns.namespacePreorder(cases[1].namespace, cases[1].salt, 96, { sender: bob });
+      expect(receipt.success).eq(true);
+      expect(receipt.result).include('u12');
+    });
+
+    it("should fail", async () => {
+      let receipt = await bns.namespaceReveal(
+        cases[1].namespace, 
+        cases[1].salt,
+        cases[1].priceFunction, 
+        cases[1].renewalRule, 
+        cases[1].nameImporter, { sender: bob });
+      expect(receipt.success).eq(false);
+      expect(receipt.error).include('1012');
+    });
+  });
+
+  describe("Given an existing, valid pre-order from Alice for the namespace 'id' initiated at block #31, revealing the namespace", () => {
+
+    beforeEach(async () => {
+      let receipt = await bns.namespacePreorder(cases[1].namespace, cases[1].salt, cases[1].value, { sender: cases[1].namespaceOwner });
+      expect(receipt.success).eq(true);
+      expect(receipt.result).include('u12');
+    });
+
+    it("should succeed if the price-function, lifetime, namespace and salt are valid", async () => {
+      let receipt = await bns.namespaceReveal(
+        cases[1].namespace, 
+        cases[1].salt,
+        cases[1].priceFunction, 
+        cases[1].renewalRule, 
+        cases[1].nameImporter, { sender: cases[1].namespaceOwner });
+      expect(receipt.success).eq(true);
+      expect(receipt.result).include('true');
+    });
+  });
```

Comment:
Can you add a test to ensure that Alice can't re-reveal the same namespace, and that if Bob has a competing preorder, he cannot reveal the namespace if Alice does so first?

Reply:
Addressed with https://github.com/blockstack/stacks-blockchain/pull/2013/commits/6420e13279e16b423247dba5cce7e93a9a695e79